lib/corpus.py

```diff
@@ -301,6 +301,154 @@ def _dump_internal(self, out: TextIO, indentation: str = ""):
             out.write("%s</subcorpus>\n" % (indentation,))
 
 
+class CorpusV2(Corpus):
+    """
+    This class represents a corpus in the Bliss format. It is also used to represent subcorpora when the parent_corpus
+    attribute is set. Corpora with include statements can be read but are written back as a single file.
+
+    The difference with respect to :class:`i6_core.lib.corpus.Corpus` is that in this class we can access
+    any recording or segment in practically O(1).
+    """
+
+    def __init__(self):
+        super().__init__()
+
+        self.parent_corpus: Optional[Corpus] = None
+
+        self.subcorpora: Dict[str, Corpus] = {}
+        self.recordings: Dict[str, RecordingV2] = {}
```

Comment:
how about we add the new Dict members under a different name and then add a property like
```
@property
def subcorpora(self):
    return list(self._subcorpora.values())

@subcorpora.setter
def subcorpora(self, values):
    for sc in values:
        self._subcorpora[sc.name] = sc
```
so that most previous code would also work with V2. Also many functions below would then not need to be re-implemented.

Reply:
In my view, we shouldn't restrain ourselves to using V1 code anymore, since it leads the user to keep using retrocompatible code instead of striving for optimal performance. If the user needs a `Corpus`, then they should build a `Corpus`, and should only build a `CorpusV2` if they know what they're doing.

If you still think that it would be a nice addition, please let me know, we can debate it.

Anyway, I expect that the `CorpusV2` class makes sense only in very specific scenarios, such as when having to access segments in O(1). Otherwise, the `Corpus` class still makes sense as a repository of subcorpora, recordings and segments, since the functionality in `V2` won't need to be used by many users.

Reply:
> In my view, we shouldn't restrain ourselves to using V1 code anymore

This would be another argument for not inheriting from Corpus. Maybe a Corpus-like implementation is not even necessary. As you said 

> I expect that the CorpusV2 class makes sense only in very specific scenarios,

Maybe for these scenarios a simple data structure or even just a 
```
segment_map = { s.fullname(): s for s in some_corpus.all_segments() }
```
would be sufficient to fulfill their performance needs.

Reply:
Yes, that makes a lot of sense. This implementation was originally thought in order to get the orthography of a given segment in O(1), which was the issue we were facing. @michelwi do you think returning a segment map from full names to segments would do the trick for your specific use case (I don't know all the details)?

@albertz and the rest of the reviewers, what's your take on this? Do you think the functionality here is too redundant? Our use case could be covered by a small function (or two, if you want to get recordings in O(1) as well) in the `Corpus` class.

Reply:
About the memory redundancy, I don't think there is really any problem. It will only have duplicated some of the pointers. You probably won't even notice this.

I haven't really checked too much, but if you can just put your extension into the `Corpus` class itself, or rewrite `Corpus` such that it stays compatible, I think I would prefer that over having a `CorpusV2`.